nntrainer/utils/profiler.cpp

```diff
@@ -0,0 +1,220 @@
+  auto iter = time_taken.find(event);
+
+  if (iter == time_taken.end() || iter->second.empty()) {
+    std::stringstream ss;
+    ss << "event has never recorded" << event_to_str(event);
+    throw std::invalid_argument("event has never recorded");
+  }
+
+  return iter->second.back();
+}
+
+void GenericProfileListener::report(std::ostream &out) const {
+  const std::vector<unsigned int> column_size = {10, 23, 23, 23};
+  auto total_col_size =
+    std::accumulate(column_size.begin(), column_size.end(), 0);
+
+  if (warmups != 0) {
+    out << "warm up: " << warmups << '\n';
+  }
+
+  /// creating header
+  // clang-format off
+  out << std::setw(column_size[0]) << "key"
+      << std::setw(column_size[1]) << "avg"
+      << std::setw(column_size[2]) << "min"
+      << std::setw(column_size[3]) << "max" << '\n';
+  // clang-format on
+
+  // seperator
+  out << std::string(total_col_size, '=') << '\n';
+
+  /// calculate metrics while skipping warmups
+  for (auto &entry : time_taken) {
+    std::chrono::milliseconds avg{0};
+    std::chrono::milliseconds min{std::chrono::milliseconds::max()};
+    std::chrono::milliseconds max{std::chrono::milliseconds::min()};
+
+    unsigned int len = entry.second.size();
+
+    if (warmups >= len) {
+      out << std::left << std::setw(total_col_size) << event_to_str(entry.first)
+          << "less data then warmup\n";
+      continue;
+    }
+
+    for (unsigned int i = warmups; i < len; i++) {
+      auto &item = entry.second[i];
+      avg += item;
+      min = std::min(min, item);
+      max = std::max(max, item);
+    }
+    avg /= len;
+
+    // clang-format off
+    out << std::setw(column_size[0]) << event_to_str(entry.first)
+        << std::setw(column_size[1]) << avg.count()
+        << std::setw(column_size[2]) << min.count()
+        << std::setw(column_size[3]) << max.count() << '\n';
+    // clang-format on
+  }
+}
+
+Profiler &Profiler::Global() {
+  static Profiler instance;
+  return instance;
+}
+
+std::string event_to_str(const int event) {
+  switch (event) {
+  case EVENT::NN_FORWARD:
+    return "nn_forward";
+  case EVENT::TEMP:
+    return "temp";
+  }
+
+  std::stringstream ss;
+  ss << "undef(" << event << ')';
+  return ss.str();
+}
+
+void Profiler::start(const int &event) {
+  /// @todo: consider race condition
+  auto iter = start_time.find(event);
+
+  if (iter != start_time.end()) {
+    throw std::invalid_argument("profiler has already started");
+  }
+
+  start_time[event] = std::chrono::steady_clock::now();
+}
+
+void Profiler::end(const int &event) {
+  /// @todo: consider race condition
+  auto end = std::chrono::steady_clock::now();
+  auto iter = start_time.find(event);
+
+  if (iter == start_time.end()) {
+    throw std::invalid_argument("profiler hasn't started with the event");
+  }
+
+  auto duration =
+    std::chrono::duration_cast<std::chrono::milliseconds>(end - iter->second);
+  notify(event, duration);
+
+  start_time.erase(iter);
```

Comment:
lets not erase, it can be expensive. It seems like this is used for ensuring that a started profiler is ended.
However, these kinds of checks are adding overhead to the profiler. 
IMHO, profiler must be as light weight as possible.

Reply:
I'll move this to #if Debug

Reply:
And additional checks